### PR TITLE
Fix #530. Migrate to Webpack 4

### LINKED ISSUE #530: Migrate to Webpack 4

### PR DESCRIPTION
**Pull request by kayandra**
Upgraded from Webpack 2 to 4. Also used the opportunity to upgrade babel.

Then, I replaced babel es2015 preset with the env preset. I added a new script option `build-analyze` which generates a report of the files that were bundled.

Finally, the new build sits just on 3kb (an insignificant difference 🤷‍♂️🤷‍♂️🤷‍♂️ )

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -22,20 +24,27 @@ module.exports = {
...
-                },
-                compress: {
-                    screw_ie8: true
-                },
-                comments: false
-            }),
-            new webpack.BannerPlugin({banner})
+            new webpack.BannerPlugin({ banner }),
+            new BundleAnalyzerPlugin({
+                analyzerMode: process.env.ANALYZE_BUIlD ? 'server' : 'disabled'
```
Comment (pedroabreu) [inline webpack.config.js]: Shouldn't it be `ANALYZE_BUILD` instead ?

Comment (kayandrae07) [reply webpack.config.js]: Can you explain?

Comment (pedroabreu) [reply webpack.config.js]: `l` is not capitalised 🤔 so it would be `ANALYZE_BUILD` instead of `ANALYZE_BUIlD`.

Makes it easier to read imo

Comment (kayandrae07) [reply webpack.config.js]: Oh, thank you. Let me fix that

File: package.json [hunk 2]
```diff
@@ -33,14 +32,22 @@
...
+    "webpack": "^4.5.0",
+    "webpack-bundle-analyzer": "^2.11.1",
+    "webpack-cli": "^2.0.14"
   },
   "scripts": {
     "build": "npm run build-debug && npm run build-min",
-    "build-debug": "webpack",
-    "build-min": "cross-env NODE_ENV=production webpack --optimize-minimize",
+    "build-debug": "webpack --mode=development",
+    "build-analyze": "cross-env ANALYZE_BUIlD=true npm run prepublish",
```
Comment (pedroabreu) [inline package.json]: Typo also here `ANALYZE_BUILD`

Good job on this one 👍 Hopefully it will be merged 🎉

Comment (kayandrae07) [reply package.json]: Thanks

File: package.json [hunk 3]
```diff
@@ -33,14 +32,22 @@
...
+    "build-analyze": "cross-env ANALYZE_BUILD=true npm run prepublish",
+    "build-min": "cross-env NODE_ENV=production webpack --mode=production",
     "build-watch": "webpack --watch",
     "test": "karma start --single-run",
     "prepublish": "npm run build"
-  }
+  },
+  "browserslist": [
+    "> 1%",
+    "last 2 versions"
```
Comment (vizjerai) [inline package.json]: May want to remove `last 2 versions`

https://jamie.build/last-2-versions

Comment (kayandrae07) [reply package.json]: @vizjerai that's an interesting read and I agree with Jamie on this

**Comment by zenorocha** [comment]
Hey @kayandrae07, I see you deleted the `dist` folder. I would like to keep it for now because people using Bower can still benefit from it.

**Comment by kayandrae07** [comment]
Hi @zenorocha the dist folder is a part of PR. Here's the PR repo https://github.com/kayandrae07/clipboard.js/tree/update-webpack-4

**Comment by Ruffio** [comment]
Can this PR be merged or are we missing something?

**Comment by Ruffio** [comment]
Can this PR be merged or are we missing something?

**Comment by kayandrae07** [comment]
@Ruffio waiting on @zenorocha

**Comment by Ruffio** [comment]
@zenorocha any comments or can this PR be merged?

**Comment by pedroabreu** [comment]
Perhaps it's best to switch to rollup as #552 . Worth the discussion as it seems that it has size gains (or loss)

**Comment by zenorocha** [comment]
This has been merged. Thanks a lot guys and sorry for taking so long.

**Pull request merged**